lazer/sdk/rust/client/src/resilient_ws_connection.rs

```diff
@@ -0,0 +1,193 @@
+};
+use tokio::{pin, select, sync::mpsc, time::Instant};
+use tracing::{error, info, warn};
+
+use crate::{
+    ws_connection::{AnyResponse, PythLazerWSConnection},
+    CHANNEL_CAPACITY,
+};
+use anyhow::{bail, Context, Result};
+
+const BACKOFF_RESET_DURATION: Duration = Duration::from_secs(10);
+
+pub struct PythLazerResilientWSConnection {
+    request_sender: mpsc::Sender<Request>,
+}
+
+impl PythLazerResilientWSConnection {
+    /// Creates a new resilient WebSocket client instance
+    ///
+    /// # Arguments
+    /// * `endpoint` - The WebSocket URL of the Lazer service
+    /// * `access_token` - Access token for authentication
+    /// * `sender` - A sender to send responses back to the client
+    ///
+    /// # Returns
+    /// Returns a new client instance (not yet connected)
+    pub fn new(
+        endpoint: String,
+        access_token: String,
+        backoff: ExponentialBackoff,
+        sender: mpsc::Sender<AnyResponse>,
+    ) -> Self {
+        let (request_sender, mut request_receiver) = mpsc::channel(CHANNEL_CAPACITY);
+        let mut task = PythLazerResilientWSConnectionTask::new(endpoint, access_token, backoff);
+
+        tokio::spawn(async move {
+            if let Err(e) = task.run(sender, &mut request_receiver).await {
+                error!("Resilient WebSocket connection task failed: {}", e);
+            }
+        });
+
+        Self { request_sender }
+    }
+
+    pub async fn subscribe(&mut self, request: SubscribeRequest) -> Result<()> {
+        self.request_sender
+            .send(Request::Subscribe(request))
+            .await
+            .context("Failed to send subscribe request")?;
+        Ok(())
+    }
+
+    pub async fn unsubscribe(&mut self, subscription_id: SubscriptionId) -> Result<()> {
+        self.request_sender
+            .send(Request::Unsubscribe(UnsubscribeRequest { subscription_id }))
+            .await
+            .context("Failed to send unsubscribe request")?;
+        Ok(())
+    }
+}
+
+struct PythLazerResilientWSConnectionTask {
+    endpoint: String,
+    access_token: String,
+    subscriptions: Vec<SubscribeRequest>,
+    backoff: ExponentialBackoff,
+}
+
+impl PythLazerResilientWSConnectionTask {
+    pub fn new(endpoint: String, access_token: String, backoff: ExponentialBackoff) -> Self {
+        Self {
+            endpoint,
+            access_token,
+            subscriptions: Vec::new(),
+            backoff,
+        }
+    }
+
+    pub async fn run(
+        &mut self,
+        response_sender: mpsc::Sender<AnyResponse>,
+        request_receiver: &mut mpsc::Receiver<Request>,
+    ) -> Result<()> {
+        let mut last_failure_time = Instant::now();
+
+        loop {
+            if let Err(e) = self.start(response_sender.clone(), request_receiver).await {
+                if last_failure_time.elapsed() > BACKOFF_RESET_DURATION {
+                    self.backoff.reset();
+                }
```

Comment:
This feels like a hack. I think we should reset the backoff object in `start()` when we successfully connect and subscribe.

Reply:
I want the behavior to be a little different than that. For example, if client can connect and subscribe, but get disconnected immediately after we should not reset.

Reply:
It only happens when `max_interval` is greater than 10 seconds. Otherwise it will try connecting every `max_interval`. 
I think the best solution is to set BACKOFF_RESET_DURATION to be 2*max_interval.

Reply:
I fixed it another way. Now we check the duration of the connection and if was connected for more than BACKOFF_RESET_DURATION and timout + second it was connected successfully and we reset the backoff.